File: src/main/java/org/kite9/diagram/visualization/compaction/segment/Segment.java
package org.kite9.diagram.visualization.compaction.segment;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.kite9.diagram.common.BiDirectional;
import org.kite9.diagram.common.algorithms.det.DetHashSet;
import org.kite9.diagram.common.algorithms.so.AlignStyle;
import org.kite9.diagram.common.algorithms.so.Slideable;
import org.kite9.diagram.common.elements.Dimension;
import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.compaction.Compaction;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.framework.common.Kite9ProcessingException;
import org.kite9.framework.logging.LogicException;


/**
 * A segment is a set of vertices that must have the same horizontal or vertical position.
 * 
 * 
 * @author robmoffat
 *
 */
public class Segment implements Comparable<Segment> {
	
	private Dimension dimension;
	private int i;
	private Slideable<Segment> slideable;
	private Set<UnderlyingInfo> underlyings;
	private Set<Rectangular> rectangulars;
	private Set<Connection> connections;
	private AlignStyle alignStyle = AlignStyle.CENTER;	// default

	public AlignStyle getAlignStyle() {
		return alignStyle;
	}

	public void setAlignStyle(AlignStyle alignStyle) {
		this.alignStyle = alignStyle;
	}

	public Slideable<Segment> getSlideable() {
		return slideable;
	}

	public void setSlideable(Slideable<Segment> slideable) {
		this.slideable = slideable;
	}
	
	public Set<UnderlyingInfo> getUnderlyingInfo() {
		if (underlyings == null) {
			underlyings = getDartsInSegment().stream().flatMap(o -> convertUnderlyingToUnderlyingInfo(o))
			.filter(a -> a.getDiagramElement() != null)
			.collect(Collectors.toSet());
		}
		
		return underlyings;
	}
	
	public Side getSingleSide() {
		if (getUnderlyingInfo().size() > 1) {
			throw new LogicException();
		} else {
			return getUnderlyingInfo().iterator().next().getSide();
		}
	}
	
	public DiagramElement getUnderlyingWithSide(Side s) {
		return getUnderlyingInfo().stream().filter(ui -> ui.getSide() == s).map(ui -> ui.getDiagramElement()).findFirst().orElse(null);
	}
	
	public boolean hasUnderlying(DiagramElement de) {
		return underlyings.stream()
				.map(u -> u.getDiagramElement())
				.filter(a -> a == de)
				.count() > 0;
	}
	
	private Stream<UnderlyingInfo> convertUnderlyingToUnderlyingInfo(Dart d) {
		Map<DiagramElement, Direction> diagramElements = d.getDiagramElements();
		return diagramElements.keySet().stream().map(de -> toUnderlyingInfo(de, diagramElements.get(de)));
	}

	private UnderlyingInfo toUnderlyingInfo(DiagramElement de, Direction d) {
		return new UnderlyingInfo(de, 
				getSideFromDirection(de, d));
	}
	
	private Side getSideFromDirection(DiagramElement de, Direction d) {
		if (de instanceof BiDirectional) {
			return Side.NEITHER;
		} else if (de instanceof Rectangular) {
			switch (d) {
			case DOWN:
			case RIGHT:
				return Side.END;
			case UP:
			case LEFT: 
				return Side.START;
			default:
				return Side.NEITHER;
			}
		} else {
			throw new Kite9ProcessingException();
		}
	}

	public Segment(Dimension dimension, int i) {
		this.dimension = dimension;
		this.i = i;
	}
	
	public void addToSegment(Vertex v) {
		verticesInSegment.add(v);
		underlyings = null;
	}

	public String getIdentifier() {
		return dimension+" ("+i+" "+getUnderlyingInfo()+" "+alignStyle+" )";
	}
	
	@Override
	public String toString() {
		 return getIdentifier() + " "+verticesInSegment.toString();
	}
	
	private Set<Vertex> verticesInSegment = new LinkedHashSet<Vertex>();

	public Set<Vertex> getVerticesInSegment() {
		return verticesInSegment;
	}
	
	public boolean connects(Vertex a, Vertex b) {
		return (inSegment(a) && inSegment(b));
	}

	private boolean inSegment(Vertex b) {
		return verticesInSegment.contains(b);
	}

	public Dimension getDimension() {
		return dimension;
	}

	/**
	 * De-facto ordering for segments.
	 */
	public int compareTo(Segment o) {
		return getSlideable().getMinimumPosition().compareTo(o.getSlideable().getMinimumPosition());
	}
	
	/**
	 * This is a utility method, used to set the positions of the darts for the diagram
	 */
	public Collection<Dart> getDartsInSegment() {
		Collection<Dart> darts = new DetHashSet<Dart>();
		for (Vertex v : verticesInSegment) {
			for (Edge e : v.getEdges()) {
				if (e instanceof Dart) { 
					if (dimension==Dimension.H) {
						if ((e.getDrawDirection()==Direction.LEFT) || (e.getDrawDirection()==Direction.RIGHT)) {
							darts.add((Dart)e);
						}
					} else if (dimension==Dimension.V) {
						if ((e.getDrawDirection()==Direction.UP) || (e.getDrawDirection()==Direction.DOWN)) {
							darts.add((Dart)e);
						}
					}
				}
			}
		}
		
		return darts;
	}

	public int getNumber() {
		return i;
	}
	

	public Set<Rectangular> getRectangulars() {
		if (rectangulars==null) {
			rectangulars = getUnderlyingInfo().stream()
					.map(ui -> ui.getDiagramElement())
					.filter(de -> de instanceof Rectangular)
					.map(de -> (Rectangular) de).collect(Collectors.toSet());
		} 
		
		return rectangulars;
	}
	
	public Set<Connection> getConnections() {
		if (connections==null) {
			connections = getUnderlyingInfo().stream()
					.map(ui -> ui.getDiagramElement())
					.filter(de -> de instanceof Connection)
					.map(de -> (Connection) de).collect(Collectors.toSet());
		} 
		
		return connections;
	}
	
	private Set<Segment> leavingSegments = null;
	
	public Set<Segment> getAdjoiningSegments(Compaction c) {
		if (leavingSegments == null) {
			boolean isHorizontal = dimension == Dimension.H;
			
			// find segments that meet this one
			leavingSegments = getVerticesInSegment().stream()
				.map(v -> isHorizontal ? c.getVerticalVertexSegmentMap().get(v) : c.getHorizontalVertexSegmentMap().get(v))
				.filter(x -> x != null)
				.collect(Collectors.toSet());
				
		}
		return leavingSegments;
	}
}

File: src/main/kotlin/org/kite9/diagram/visualization/compaction/segment/Segment.kt
package org.kite9.diagram.visualization.compaction.segment

import org.kite9.diagram.common.BiDirectional
import org.kite9.diagram.common.algorithms.so.AlignStyle
import org.kite9.diagram.common.algorithms.so.Slideable
import org.kite9.diagram.common.elements.Dimension
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.Rectangular
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.orthogonalization.Dart
import java.util.stream.Collectors

import org.kite9.diagram.visualization.compaction.Compaction
import java.util.function.Predicate


/**
 * A segment is a set of vertices that must have the same horizontal or vertical position.
 *
 *
 * @author robmoffat
 */
class Segment(val dimension: Dimension, val number: Int) : Comparable<Segment> {

    var slideable: Slideable<Segment>? = null

    val bob: Int by lazy { 6 }

    val rectangulars: Set<Rectangular> by lazy {
        underlyingInfo
            .map { it.diagramElement }
            .filterIsInstance<Rectangular>()
            .toSet()
        }

    val connections: Set<Connection> by lazy {
        underlyingInfo
            .map { it.diagramElement }
            .filter { it is Connection }
            .map { it as Connection }
            .toSet()
    }

    var alignStyle : AlignStyle? = AlignStyle.CENTER // default

    val underlyingInfo: Set<UnderlyingInfo> by lazy {
        dartsInSegment
            .flatMap { convertUnderlyingToUnderlyingInfo(it) }
            .filter { a: UnderlyingInfo -> a.diagramElement != null }
            .toSet()
    }


    var singleSide: Side? = null
        get() {
            if (field == null) {
                field = underlyingInfo!!.stream()
                    .map { ui: UnderlyingInfo -> ui.side }
                    .reduce(null) { a: Side?, b: Side? -> sideReduce(a, b) }
            }
            return field
        }
        private set

    private fun sideReduce(a: Side?, b: Side?): Side? {
        return if (a == null) {
            b
        } else if (b == null) {
            a
        } else if (a == b) {
            a
        } else {
            Side.BOTH
        }
    }

    fun getUnderlyingWithSide(s: Side): DiagramElement? {
        return underlyingInfo
            .filter { it.side == s }
            .map { it.diagramElement }
            .firstOrNull()
    }

    fun hasUnderlying(de: DiagramElement): Boolean {
        return underlyingInfo
            .map { u: UnderlyingInfo -> u.diagramElement }
            .filter { a: DiagramElement -> a === de }
            .count() > 0
    }

    fun hasUnderlying(des: Set<DiagramElement>): Boolean {
        return underlyingInfo
            .map { u: UnderlyingInfo -> u.diagramElement }
            .filter { a: DiagramElement -> des.contains(a) }
            .count() > 0
    }

    private fun convertUnderlyingToUnderlyingInfo(d: Dart): Iterable<UnderlyingInfo> {
        val diagramElements: Map<DiagramElement, Direction> = d.getDiagramElements()
        return diagramElements.keys.map { toUnderlyingInfo(
            it,
            diagramElements[it])
        }
    }

    private fun toUnderlyingInfo(de: DiagramElement, d: Direction?): UnderlyingInfo {
        return UnderlyingInfo(
            de,
            getSideFromDirection(de, d)
        )
    }

    private fun getSideFromDirection(de: DiagramElement, d: Direction?): Side {
        return if (de is BiDirectional<*>) {
            Side.NEITHER
        } else if (de is Rectangular) {
            when (d) {
                Direction.DOWN, Direction.RIGHT -> Side.END
                Direction.UP, Direction.LEFT -> Side.START
                else -> Side.NEITHER
            }
        } else {
            throw LogicException()
        }
    }

    fun addToSegment(v: Vertex) {
        verticesInSegment.add(v)
    }

    val identifier: String
        get() = "$dimension ($number $underlyingInfo $alignStyle )"

    override fun toString(): String {
        return "$identifier $verticesInSegment"
    }

    private val verticesInSegment: MutableSet<Vertex> = LinkedHashSet()

    fun getVerticesInSegment(): Set<Vertex> {
        return verticesInSegment
    }

    fun connects(a: Vertex, b: Vertex): Boolean {
        return inSegment(a) && inSegment(b)
    }

    private fun inSegment(b: Vertex): Boolean {
        return verticesInSegment.contains(b)
    }

    /**
     * De-facto ordering for segments.
     */
    override fun compareTo(o: Segment): Int {
        return slideable!!.minimumPosition.compareTo(o.slideable!!.minimumPosition)
    }

    /**
     * This is a utility method, used to set the positions of the darts for the diagram
     */
    val dartsInSegment: Collection<Dart> by lazy {
        verticesInSegment
            .flatMap { it.getEdges() }
            .filterIsInstance<Dart>()
            .filter {
                if (dimension === Dimension.H) {
                    (it.getDrawDirection() === Direction.LEFT || it.getDrawDirection() === Direction.RIGHT)
                } else {
                    (it.getDrawDirection() === Direction.UP || it.getDrawDirection() === Direction.DOWN)
                }
            }
    }

    /**
     * Used for deciding whether a segment should be pushed left/right/up/down according to it's connections.
     */
    val adjoiningSegmentBalance: Int
        get() {
            val isHorizontal = dimension === Dimension.H
            return getVerticesInSegment().stream()
                .mapToInt { v: Vertex ->
                    if (isHorizontal) dartCount(v, Direction.DOWN) - dartCount(
                        v,
                        Direction.UP
                    ) else dartCount(v, Direction.RIGHT) - dartCount(v, Direction.LEFT)
                }
                .sum()
        }

    private fun dartCount(v: Vertex, d: Direction): Int {
        for (e in v.getEdges()) {
            if (e is Dart && e.getDrawDirectionFrom(v) === d) {
                return 1
            }
        }
        return 0
    }


    private var leavingSegments: Set<Segment>? = null

    fun getAdjoiningSegments(c: Compaction): Set<Segment?>? {
        if (leavingSegments == null) {
            val isHorizontal = dimension === Dimension.H

            // find segments that meet this one
            leavingSegments = getVerticesInSegment()
                .map {
                    if (isHorizontal) c.verticalVertexSegmentMap[it] else c.horizontalVertexSegmentMap[it]
                }
                .filterNotNull()
                .toSet()
        }

        return leavingSegments
    }

}